lib/Doctrine/ORM/Tools/SchemaTool.php

```diff
@@ -910,4 +911,42 @@ public function getUpdateSchemaSql(array $classes, $saveMode = false)
 
         return $schemaDiff->toSql($this->platform);
     }
+
+    /**
+     * Creates the schema from the database, ensuring tables from the target schema are whitelisted for comparison.
+     */
+    private function createSchemaForComparison(Schema $toSchema) : Schema
+    {
+        // check and memorize if we are using dbal >= 3.0
+        static $hasFilterExpressionMethod;
+        $hasFilterExpressionMethod = $hasFilterExpressionMethod ?? method_exists(Configuration::class, 'getFilterSchemaAssetsExpression');
+
+        // backup schema assets filter and filter-expression
+        $sm               = $this->em->getConnection()->getSchemaManager();
+        $config           = $this->em->getConnection()->getConfiguration();
+        $filterExpression = $hasFilterExpressionMethod ? $config->getFilterSchemaAssetsExpression() : null;
```

Comment:
This code is not needed as `filterExpressions` are deprecated and are transalted to the new assets filter internally, and since you also have the old `$filter` there and set it anyways, this shouldn't be needed.

Reply:
The check is required to me as that's the only way to get the state of the config object when considering legacy and non legacy usages.

Reply:
Wait... @beberlei seems right, though. ORM v2.7 requires DBAL 2.9, which is the version that deprecated these methods and introduced the new ones...

@nicolas-grekas can you explain a bit more why the check is required to you?

I mean, even though `setSchemaAssetsFilter()` resets the filter expression, the callable is the source of truth so this seems unnecessary to me. Perhaps @morozov would like to weight in too?

Reply:
When `setSchemaAssetsFilter()` is called to setup the decorating closure, we change the state of the config object.
We then have to restore this object to the same exact state it was before the call in the `finally` block.
And there is no other way to do so that calling `setFilterSchemaAssetsExpression` or `setSchemaAssetsFilter` conditionally as done in the PR...

Reply:
Right, that's what I meant with my comment. But the expression is irrelevant since DBAL takes care of converting it into a callable, which is what's being used to revert stuff...

What I'm trying to say here is that we're trying to manage DBAL internals in the ORM and that's not required.

Reply:
Not sure to get teh reasoning, but it doesn't matter :)
Updated.

Reply:
But note that we had tests checking for the situation.
Please advise if I should remove them.

Reply:
@nicolas-grekas don't worry. I'll check the tests and see what can be cleaned up there. Thanks for updating things here.